src/librustc_typeck/check/compare_method.rs

```diff
@@ -138,14 +138,51 @@ pub fn compare_impl_method<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
     }
 
     if impl_m.fty.sig.0.inputs.len() != trait_m.fty.sig.0.inputs.len() {
-        span_err!(tcx.sess, impl_m_span, E0050,
+        let trait_number_args = trait_m.fty.sig.0.inputs.len();
+        let impl_number_args = impl_m.fty.sig.0.inputs.len();
+        let impl_m_node_id = tcx.map.as_local_node_id(impl_m.def_id).unwrap();
+        let mspan = match tcx.map.expect_impl_item(impl_m_node_id).node {
+            ImplItemKind::Method(ref impl_m_sig, _) => {
+                if let Some(arg) = impl_m_sig.decl.inputs.get(
+                    if trait_number_args > 0 {
+                        trait_number_args - 1
+                    } else {
+                        0
+                    }) {
+                    arg.pat.span
+                } else {
+                    impl_m_span
+                }
+            }
+            _ => bug!("{:?} is not a method", impl_m)
+        };
+        let span = impl_m_span;
+        struct_span_err!(tcx.sess, mspan, E0050,
             "method `{}` has {} parameter{} \
              but the declaration in trait `{}` has {}",
             trait_m.name,
-            impl_m.fty.sig.0.inputs.len(),
-            if impl_m.fty.sig.0.inputs.len() == 1 {""} else {"s"},
+            impl_number_args,
+            if impl_number_args == 1 {""} else {"s"},
             tcx.item_path_str(trait_m.def_id),
-            trait_m.fty.sig.0.inputs.len());
+            trait_number_args)
+            .span_label(trait_item_span.unwrap(),
+                        &format!("trait requires {}",
+                                 &if trait_number_args != 1 {
+                                     format!("{} parameters", trait_number_args)
+                                 } else {
+                                     format!("{} parameter", trait_number_args)
+                                 }))
+            .span_label(span,
```

Comment:
Played around a bit more.  You can see what I ended up doing here: https://github.com/jonathandturner/rust/tree/e0050

Changes:
* Focused the trait secondary span on the parameter rather than underlining everything
* Focused the impl primary span on the parameter rather than underlining everything in the case of the trait having more parameters than the impl
* Made the displaying of the trait span optional since the trait may have come from outside the current crate